### DISCUSSION TITLE
addFrame by combining two existing frames

### DISCUSSION THREAD

**Discussion by arjung128** [stack-of-tasks/pinocchio#1927] [Q&A] [answered]
Hello, I would like to create a new frame based on two existing frames. Namely, `joint_gripper_fingertip_right` and `joint_gripper_fingertip_left` exist, and I would like to keep track of the center point between these two frames, i.e. the average position of the fingertips. 

I know the following is the syntax:

```
frame = pin.Frame(
    name="joint_gripper_fingertip_center",
    parent_joint=2,
    parent_frame=6,
    placement=pin.SE3(
        rotation=np.eye(3),
        translation=np.array([0, 0, 0]),
    ),
    type=pin.FrameType.BODY,
)
model.addFrame(frame)
```

What I am unsure about is what the `parent_joint` and `parent_frame` in this situation would be. Also, how can the `placement` be made to be the average of the positions of the left and right gripper fingertips instead of a hardcoded pose? And `pin.FrameType.BODY` would be fine as the `type` for this, correct?

I'm sorry if this is a silly question. I tried looking through the other issues as well as the examples but couldn't quite find anything to help me beyond this point. Any help will be much appreciated, thanks in advance!

**Comment by jcarpent** [comment, marked as answer]
> What I am unsure about is what the `parent_joint` and `parent_frame` in this situation would be. 

`parent_frame` attribute is there just for the record of the underlying tree of frames.
`parent_joint` is the quantity you should take care

> And pin.FrameType.BODY would be fine as the type for this, correct?

You should rather use `OP_FRAME` which stands for Operational Frame, which seems to be aligned with your purposes as you are considering a virtual, non-physical, point. 

> Also, how can the placement be made to be the average of the positions of the left and right gripper fingertips instead of a hardcoded pose?

The fact that your two joints are moving independently (`joint_gripper_fingertip_right` and `joint_gripper_fingertip_left`), you should consider updating the `placement` quantity of your new frame after each change of configuration, as your averaged placement will change according to the relative placement of the two considered frames.

**Reply by arjung128** [reply]
Thanks so much for your prompt response!

> `parent_joint` is the quantity you should take care

Would the `parent_joint` be the same as that of `joint_gripper_fingertip_right`? Here is the description for `joint_gripper_fingertip_right` from the URDF:

```
<joint name="joint_gripper_fingertip_right" type="fixed">
  <origin xyz="-0.190596948563868 -0.015 0.0" rpy="-1.57079632679483 -3.43320051448326e-14 0.540456056432235"/>
  <axis xyz="0.0 0.0 0.0"/>
  <parent link="link_gripper_finger_right"/>
  <child link="link_gripper_fingertip_right"/>
</joint>
```

Would I use this information to set `parent_joint`? Or what should I set the `parent_joint` to, and where can I get this information? Apologies, I'm still unsure about what this quantity should be set to.

Once again, thanks a lot for your work and for your help!

**Reply by jcarpent** [reply]
Yes, you can use as parent joint one of the two joints already identified.

**Reply by arjung128** [reply]
Got it, thanks!

Here is my code:

```
model, geom_model, visual_model = pin.buildModelsFromUrdf('stretch/stretch-re2-2025/exported_urdf/stretch.urdf', 'stretch/stretch-re2-2025/exported_urdf/')

frame = pin.Frame(
    name="joint_gripper_fingertip_center",
    parent_joint=11, # model.parents[-2] = model.parents[-3] = 11 (parents of the right and left fingertip joints)
    parent_frame=81, # model.frames[81].name = 'joint_gripper_fingertip_right'
    placement=pin.SE3(
        rotation=np.eye(3),
        translation=np.array([0, 0, 0]),
    ),
    type=pin.FrameType.OP_FRAME,
)
model.addFrame(frame)

data, geom_data, visual_data = pin.createDatas(model, geom_model, visual_model)

q = pin.randomConfiguration(model)
pin.forwardKinematics(model, data, q)
pin.framesForwardKinematics(model, data, q)

# set the translation for joint_gripper_fingertip_center to be the average of joint_gripper_fingertip_right and joint_gripper_fingertip_left
data.oMf[model.getFrameId('joint_gripper_fingertip_center')].translation = (data.oMf[model.getFrameId('joint_gripper_fingertip_right')].translation + data.oMf[model.getFrameId('joint_gripper_fingertip_left')].translation) / 2
```

However, this doesn't seem to affect `pin.computeFrameJacobian(model,data,q,model.getFrameId('joint_gripper_fingertip_center'))`, and calling `pin.updateFramePlacements(model, data)` reverts the `joint_gripper_fingertip_center` frame to a completely different pose. Any ideas what I'm missing to get an accurate Jacobian (using pin.computeFrameJacobian) and frame placement (using pin.updateFramePlacements) after manually setting the translation to be the midpoint between the two finger joints?
